src/FrequencyBalancerAgent.cpp

```diff
@@ -0,0 +1,644 @@
+                m_last_epoch_count = epoch_count;
+                m_epoch_wait_count = MINIMUM_EPOCHS_FOR_NEW_EPOCH_CONTROL;
+                m_handle_new_epoch = true;
+            }
+        }
+        else if (std::isnan(m_last_epoch_count)) {
+            // We don't want to update the previous epoch count if the wait
+            // period has not been exceeded. BUT we must update it if the
+            // previous count was NaN (e.g., we are observing the first epoch
+            // now)
+            m_last_epoch_count = epoch_count;
+        }
+    }
+
+    void FrequencyBalancerAgent::wait(void)
+    {
+        m_waiter->wait();
+    }
+
+    std::vector<std::string> FrequencyBalancerAgent::policy_names(void)
+    {
+        return { "POWER_PACKAGE_LIMIT_TOTAL",
+                 "USE_FREQUENCY_LIMITS", "USE_SST_TF" };
+    }
+
+    std::vector<std::string> FrequencyBalancerAgent::sample_names(void)
+    {
+        return {};
+    }
+
+    std::vector<std::pair<std::string, std::string> >
+        FrequencyBalancerAgent::report_header(void) const
+    {
+        return {
+            {"Agent uses frequency control", std::to_string(m_policy_use_frequency_limits)},
+            {"Agent uses SST-TF", std::to_string(m_use_sst_tf)},
+        };
+    }
+
+    std::vector<std::pair<std::string, std::string> >
+        FrequencyBalancerAgent::report_host(void) const
+    {
+        return {};
+    }
+
+    std::map<uint64_t, std::vector<std::pair<std::string, std::string> > >
+        FrequencyBalancerAgent::report_region(void) const
+    {
+        return {};
+    }
+
+    std::vector<std::string> FrequencyBalancerAgent::trace_names(void) const
+    {
+        std::vector<std::string> names;
+        for (int package_idx = 0; package_idx < m_package_count; ++package_idx) {
+            for (auto core_idx : m_package_core_indices[package_idx]) {
+                names.push_back(std::string("NON_NET_TIME_PER_EPOCH-core-") + std::to_string(core_idx));
+            }
+        }
+        for (int package_idx = 0; package_idx < m_package_count; ++package_idx) {
+            names.push_back(std::string("DESIRED_NON_NETWORK_TIME-package-") + std::to_string(package_idx));
+        }
+        return names;
+    }
+
+    std::vector<std::function<std::string(double)> >
+        FrequencyBalancerAgent::trace_formats(void) const
+    {
+        std::vector<std::function<std::string(double)> > formats;
+        for (int package_idx = 0; package_idx < m_package_count; ++package_idx) {
+            for (size_t i = 0; i < m_package_core_indices[package_idx].size(); ++i) {
+                formats.push_back(string_format_double);
+            }
+        }
+        for (int package_idx = 0; package_idx < m_package_count; ++package_idx) {
+            formats.push_back(string_format_double);
+        }
+        return formats;
+    }
+
+    void FrequencyBalancerAgent::trace_values(std::vector<double> &values)
+    {
+        values.clear();
```

Comment:
Seems like this should have been taken care of by the caller (i.e. in the Controller during walk_up).

This way of updating ```values``` also differs slightly from how other agents (PowerBalancerAgent, PowerGovernorAgent, FFNetAgent) do things.  They all assume the vector is fixed, and overwrite all the values during each call to this method.